File: src/libs/MULTISIG.ts
import { PrivateKey, PublicKey } from '@bsv/sdk/primitives';
import Script from '@bsv/sdk/script/Script';
import Transaction from '@bsv/sdk/transaction/Transaction';
import OP from '@bsv/sdk/script/OP';
// 引入需要的工具类
import { hash256, sha256 } from '@bsv/sdk/primitives/Hash';
import type ScriptChunk from '@bsv/sdk/script/ScriptChunk';
import type ScriptTemplate from '@bsv/sdk/script/ScriptTemplate';
import type ScriptTemplateUnlock from '@bsv/sdk/script/ScriptTemplateUnlock';
import LockingScript from '@bsv/sdk/script/LockingScript';
import UnlockingScript from '@bsv/sdk/script/UnlockingScript';
import { toHex } from '@bsv/sdk/primitives/utils';
import TransactionSignature from '@bsv/sdk/primitives/TransactionSignature';

// 错误定义
const Errors = {
  INVALID_PUBLIC_KEYS: new Error('invalid public keys'),
  NO_PRIVATE_KEYS: new Error('private keys not supplied'),
  INVALID_M: new Error('invalid m value'),
  EMPTY_PREVIOUS_TX: new Error('empty previous tx')
};

/**
 * MultiSig 表示 M-of-N 多重签名模板
 */
export default class MultiSig implements ScriptTemplate {
  // private privateKeys: PrivateKey[];
  // public publicKeys: PublicKey[];
  // public m: number; // 需要的签名数量
  // public n: number; // 公钥总数
  // public sigHashType: number;

  /**
   * 创建一个用于解锁的 MultiSig 实例
   * @param privateKeys 用于签名的私钥数组
   * @param publicKeys 公钥数组
   * @param m 需要的签名数量 (m-of-n)
   * @param sigHashType SIGHASH 类型 (默认: SIGHASH_ALL | SIGHASH_FORKID = 0x41)
   */
  // constructor(
  //   privateKeys: PrivateKey[],
  //   publicKeys: PublicKey[],
  //   m: number,
  //   sigHashType: number = 0x41 // SIGHASH_ALL | SIGHASH_FORKID
  // ) {
  //   if (publicKeys.length === 0 || publicKeys.length > 20) {
  //     throw Errors.INVALID_PUBLIC_KEYS;
  //   }
  //   if (m <= 0 || m > publicKeys.length) {
  //     throw Errors.INVALID_M;
  //   }

  //   this.privateKeys = privateKeys;
  //   this.publicKeys = publicKeys;
  //   this.m = m;
  //   this.n = publicKeys.length;
  //   this.sigHashType = sigHashType;
  // }

  /**
   * 创建 P2MS (Pay to Multi-Signature) 锁定脚本
   * @param publicKeys 公钥数组
   * @param m 需要的签名数量 (m-of-n)
   */
  // public static createLockingScript(publicKeys: PublicKey[], m: number): Script {
  //   const n = publicKeys.length;
  //   if (m <= 0 || m > n) {
  //     throw Errors.INVALID_M;
  //   }
  //   if (n === 0 || n > 20) {
  //     throw Errors.INVALID_PUBLIC_KEYS;
  //   }

  //   // 创建空脚本并添加 OP_m
  //   const script = new Script([]);
  //   script.writeOpCode(OP.OP_1 + m - 1); // OP_1 + (m-1) = OP_m

  //   // 添加公钥
  //   for (const pubKey of publicKeys) {
  //     script.writeBin(pubKey.toDER() as number[]);
  //   }

  //   // 添加 N 值和 CHECKMULTISIG
  //   script.writeOpCode(OP.OP_1 + n - 1) // OP_1 + (n-1) = OP_n
  //     .writeOpCode(OP.OP_CHECKMULTISIG);

  //   return script;
  // }

  /**
   * 实现 ScriptTemplate 接口的 lock 方法
   * 创建锁定脚本
   * @param publicKeys 公钥数组
   * @param m 需要的签名数量 (m-of-n)
   */
  public lock(publicKeys: PublicKey[], m: number): LockingScript {
    // 直接在 lock 方法中创建多签锁定脚本，而不调用 createLockingScript
    const n = publicKeys.length;
    if (m <= 0 || m > n) {
      throw Errors.INVALID_M;
    }
    if (n === 0 || n > 20) {
      throw Errors.INVALID_PUBLIC_KEYS;
    }

    // 创建空脚本并添加 OP_m
    const script = new Script([]);
    script.writeOpCode(OP.OP_1 + m - 1); // OP_1 + (m-1) = OP_m

    // 添加公钥
    for (const pubKey of publicKeys) {
      script.writeBin(pubKey.toDER() as number[]);
    }

    // 添加 N 值和 CHECKMULTISIG
    script.writeOpCode(OP.OP_1 + n - 1) // OP_1 + (n-1) = OP_n
      .writeOpCode(OP.OP_CHECKMULTISIG);

    // 创建锁定脚本对象
    const lockingScript = new LockingScript();
    lockingScript.chunks = script.chunks;
    return lockingScript;
  }

  /**
   * 实现 ScriptTemplate 接口的 unlock 方法
   * 创建一个用于解锁的 ScriptTemplateUnlock 对象
   * @param privateKeys 用于签名的私钥数组
   * @param publicKeys 公钥数组
   * @param m 需要的签名数量 (m-of-n)
   * @param sigHashType SIGHASH 类型 (默认: SIGHASH_ALL | SIGHASH_FORKID = 0x41)
   */
  public unlock(privateKeys: PrivateKey[], publicKeys: PublicKey[], m: number, sigHashType: number = 0x41): ScriptTemplateUnlock {
    // const multiSig = new MultiSig(privateKeys, publicKeys, m, sigHashType);
    
    return {
      sign: async (tx: Transaction, inputIndex: number): Promise<UnlockingScript> => {
        const script = this.sign(tx, inputIndex, privateKeys, publicKeys, m, sigHashType);
        return new UnlockingScript(script.toBinary() as unknown as ScriptChunk[]);
      },
      estimateLength: async (tx: Transaction, inputIndex: number): Promise<number> => {
        return this.estimateLength(m);
      },
    };
  }

  /**
   * 创建一个用于解锁的 MultiSig 实例
   * @param privateKeys 用于签名的私钥数组
   * @param publicKeys 公钥数组
   * @param m 需要的签名数量 (m-of-n)
   * @param sigHashType SIGHASH 类型 (默认: SIGHASH_ALL | SIGHASH_FORKID = 0x41)
   */
  // public static createUnlocker(
  //   privateKeys: PrivateKey[],
  //   publicKeys: PublicKey[],
  //   m: number,
  //   sigHashType: number = 0x41 // SIGHASH_ALL | SIGHASH_FORKID
  // ): MultiSig {
  //   // 注释掉的检查，与原始 Go 代码保持一致
  //   // if (privateKeys.length < m) {
  //   //   throw Errors.NO_PRIVATE_KEYS;
  //   // }

  //   return new MultiSig(privateKeys, publicKeys, m, sigHashType);
  // }

  /**
   * 为多重签名创建解锁脚本
   * @param tx 要签名的交易
   * @param inputIndex 要签名的输入索引
   */
  public sign(tx: Transaction, inputIndex: number, privateKeys: PrivateKey[], publicKeys: PublicKey[], m: number, sigHashType: number = 0x41): Script {
    // 检查输入的源交易输出是否存在
    if (!tx.inputs[inputIndex].sourceTXID) {
      throw Errors.EMPTY_PREVIOUS_TX;
    }

    // 获取输入的源交易输出信息
    
    
    // 创建解锁脚本
    const script = new Script([]);
    script.writeOpCode(OP.OP_0); // OP_0 for CHECKMULTISIG bug

    // 使用所需数量的私钥进行签名
    for (let i = 0; i < m; i++) {
      const input = tx.inputs[i];
      const sourceSatoshis = input.sourceTransaction?.outputs[input.sourceOutputIndex]?.satoshis;
      // 创建多签锁定脚本
      const lockingScript = input.sourceTransaction?.outputs[input.sourceOutputIndex]?.lockingScript;
      // 使用 TransactionSignature.format 计算签名哈希
      const sighashData = TransactionSignature.format({
        sourceTXID: input.sourceTXID || '',
        sourceOutputIndex: input.sourceOutputIndex,
        sourceSatoshis: sourceSatoshis as number,
        transactionVersion: tx.version,
        otherInputs: tx.inputs.filter((_, idx) => idx !== inputIndex),
        outputs: tx.outputs,
        inputIndex: inputIndex,
        subscript: lockingScript as Script,
        inputSequence: input.sequence || 0xffffffff,
        lockTime: tx.lockTime,
        scope: sigHashType
      });
      
      // 对 sighashData 进行双重哈希 (Sha256d)，然后签名
      const sig = privateKeys[i].sign(hash256(sighashData));
      
      // 添加带有 SIGHASH 类型的签名
      const signature = Buffer.concat([
        Buffer.from(sig.toDER()),
        Buffer.from([sigHashType])
      ]);
      
      // 将签名数据写入脚本
      script.writeBin([...new Uint8Array(signature)]);
    }

    return script;
  }

  /**
   * 为多重签名创建单个签名（部分签名）
   * @param tx 要签名的交易
   * @param inputIndex 要签名的输入索引
   * @param privateKey 用于签名的私钥
   */
  public signOne(tx: Transaction, inputIndex: number, privateKey: PrivateKey, sigHashType: number = 0x41): Buffer {
    // 检查输入的源交易输出是否存在
    if (!tx.inputs[inputIndex].sourceTXID) {
      throw Errors.EMPTY_PREVIOUS_TX;
    }

    // 获取输入的源交易输出信息
    const input = tx.inputs[inputIndex];
    const sourceSatoshis = input.sourceTransaction?.outputs[input.sourceOutputIndex]?.satoshis;
    
    // 创建多签锁定脚本 - 这里需要获取正确的锁定脚本
    // 假设已经设置在 input.unlockingScript 中
    const lockingScript = input.sourceTransaction?.outputs[input.sourceOutputIndex]?.lockingScript;
    
    if (!lockingScript) {
      throw new Error('锁定脚本未设置，无法生成签名');
    }
    
    // 确保 sourceSatoshis 不为 undefined
    if (sourceSatoshis === undefined) {
      throw new Error('源交易输出的 satoshis 值未设置，无法生成签名');
    }
    
    // 使用 TransactionSignature.format 计算签名哈希
    const params = {
      sourceTXID: input.sourceTXID || '',
      sourceOutputIndex: input.sourceOutputIndex,
      sourceSatoshis: sourceSatoshis as number, // 类型断言为 number，因为我们已经检查过它不是 undefined
      transactionVersion: tx.version,
      otherInputs: tx.inputs.filter((_, idx) => idx !== inputIndex),
      outputs: tx.outputs,
      inputIndex: inputIndex,
      subscript: lockingScript,
      inputSequence: input.sequence as number,
      lockTime: tx.lockTime,
      scope: sigHashType
    };
    
    // 打印所有参数
    // console.log('TransactionSignature.format 参数:');
    // console.log('sourceTXID:', params.sourceTXID);
    // console.log('sourceOutputIndex:', params.sourceOutputIndex);
    // console.log('sourceSatoshis:', params.sourceSatoshis);
    // console.log('transactionVersion:', params.transactionVersion);
    // console.log('otherInputs:', JSON.stringify(params.otherInputs));
    // console.log('outputs:', JSON.stringify(params.outputs));
    // console.log('inputIndex:', params.inputIndex);
    // console.log('subscript:', params.subscript.toHex());
    // console.log('inputSequence:', params.inputSequence);
    // console.log('lockTime:', params.lockTime);
    // console.log('scope:', params.scope);
    
    const sighashData = TransactionSignature.format(params);
    
    // console.log('=====> sighashData:', toHex(sighashData));
    // 对 sighashData 进行双重哈希 (Sha256d)，然后签名
    const sighashDataHash256 = sha256(sighashData);
    // // console.log('=====> sighashDataHash256:', toHex(sighashDataHash256));
    const sig = privateKey.sign(sighashDataHash256);
    // console.log('=====> public key:', privateKey.toPublicKey().toDER("hex"));
    
    // 返回带有 SIGHASH 类型的签名
    const signature = Buffer.concat([
      Buffer.from(sig.toDER()),
      Buffer.from([sigHashType])
    ]);
    // console.log('=====> signature:', signature.toString('hex'));
    return signature;
  }

  /**
   * 估算解锁脚本的大小（字节）
   */
  public estimateLength(m: number): number {
    // OP_0 + M * (signature + sighash flag)
    return 1 + m * (71 + 1); // 71 bytes for DER signature + 1 byte for SIGHASH
  }

  /**
   * 创建假签名脚本，用于估算大小
   * @param m 需要的签名数量
   */
  public static createFakeSign(m: number): Script {
    // 创建空脚本
    const script = new Script([]);
    script.writeOpCode(OP.OP_0);
    
    // 添加假签名（71字节每个 + 1字节 SIGHASH）
    for (let i = 0; i < m; i++) {
      // 假签名数据：使用71字节的空数据模拟签名（最大DER签名长度）+ 1字节的空SigHashFlag
      const fakeSig = Buffer.alloc(71);
      const fakeSigWithType = Buffer.concat([fakeSig, Buffer.from([0])]);
      script.writeBin([...new Uint8Array(fakeSigWithType)]);
    }
    
    return script;
  }

  /**
   * 从多个签名构建脚本
   * @param signatures 签名缓冲区数组
   */
  public static buildSignScript(signatures: Buffer[]): Script {
    // 创建空脚本
    const script = new Script([]);
    script.writeOpCode(OP.OP_0);
    
    for (const sig of signatures) {
      script.writeBin([...new Uint8Array(sig)]);
    }
    
    return script;
  }

  /**
   * 获取此多重签名的锁定脚本
   */
  // public getLockingScript(): Script {
  //   return MultiSig.createLockingScript(this.publicKeys, this.m);
  // }
}

File: pkg/libs/multisig.go
package libs

import (
	"errors"
	"fmt"

	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	script "github.com/bsv-blockchain/go-sdk/script"
	transaction "github.com/bsv-blockchain/go-sdk/transaction"
	sighash "github.com/bsv-blockchain/go-sdk/transaction/sighash"
)

var (
	ErrInvalidPublicKeys = errors.New("invalid public keys")
	ErrNoPrivateKeys     = errors.New("private keys not supplied")
	ErrInvalidM          = errors.New("invalid m value")
)

// MultiSig represents an M-of-N multisig template
type MultiSig struct {
	PrivateKeys []*ec.PrivateKey
	PublicKeys  []*ec.PublicKey
	M           int // Required signatures
	N           int // Total number of public keys
	SigHashFlag *sighash.Flag
}

// Lock creates a P2MS (Pay to Multi-Signature) locking script
func Lock(pubKeys []*ec.PublicKey, m int) (*script.Script, error) {
	n := len(pubKeys)
	if m <= 0 || m > n {
		return nil, ErrInvalidM
	}
	if n == 0 || n > 20 {
		return nil, ErrInvalidPublicKeys
	}

	s := script.NewFromBytes([]byte{})

	// Add M value
	s.AppendOpcodes(byte(script.Op1 + byte(m-1)))

	// Add public keys
	for _, pubKey := range pubKeys {
		if err := s.AppendPushData(pubKey.Compressed()); err != nil {
			return nil, err
		}
	}

	// Add N value and CHECKMULTISIG
	s.AppendOpcodes(byte(script.Op1+byte(n-1)), script.OpCHECKMULTISIG)

	return s, nil
}

// Unlock creates a new MultiSig unlocking instance
func Unlock(privKeys []*ec.PrivateKey, pubKeys []*ec.PublicKey, m int, sigHashFlag *sighash.Flag) (*MultiSig, error) {
	// if len(privKeys) < m {
	// 	fmt.Printf("len(privKeys) < m: %d < %d\n", len(privKeys), m)
	// 	return nil, ErrNoPrivateKeys
	// }

	if sigHashFlag == nil {
		shf := sighash.AllForkID
		sigHashFlag = &shf
	}

	return &MultiSig{
		PrivateKeys: privKeys,
		PublicKeys:  pubKeys,
		M:           m,
		N:           len(pubKeys),
		SigHashFlag: sigHashFlag,
	}, nil
}

// Sign creates an unlocking script for the multisig
func (ms *MultiSig) Sign(tx *transaction.Transaction, inputIndex uint32) (*script.Script, error) {
	if tx.Inputs[inputIndex].SourceTxOutput() == nil {
		return nil, transaction.ErrEmptyPreviousTx
	}

	// Print raw sighash preimage for debugging (before double SHA256)
	preimage, _ := tx.CalcInputPreimage(inputIndex, *ms.SigHashFlag)
	fmt.Printf("Go sighashData: %x\n", preimage)
	// Extract and print intermediate hashes
	if len(preimage) >= 68 {
		hashPrevouts := preimage[4:36]
		hashSequence := preimage[36:68]
		// hashOutputs is located 40 bytes from the end (4 locktime + 4 sighash flag + 32 hashOutputs)
		hashOutputs := preimage[len(preimage)-40-32 : len(preimage)-40]
		fmt.Printf("Go hashPrevouts: %x\n", hashPrevouts)
		fmt.Printf("Go hashSequence: %x\n", hashSequence)
		fmt.Printf("Go hashOutputs : %x\n", hashOutputs)
	}

	sh, err := tx.CalcInputSignatureHash(inputIndex, *ms.SigHashFlag)
	fmt.Printf("Go sighash: %x\n", sh)
	if err != nil {
		return nil, err
	}

	// Create unlocking script
	s := script.NewFromBytes([]byte{})

	// Add OP_0 for the bug in CHECKMULTISIG
	s.AppendOpcodes(script.Op0)

	// Sign with required number of private keys
	for i := 0; i < ms.M; i++ {
		sig, err := ms.PrivateKeys[i].Sign(sh)
		if err != nil {
			return nil, err
		}

		sigBuf := make([]byte, 0)
		sigBuf = append(sigBuf, sig.Serialize()...)
		sigBuf = append(sigBuf, uint8(*ms.SigHashFlag))

		if err = s.AppendPushData(sigBuf); err != nil {
			return nil, err
		}
	}

	return s, nil
}

// 分别签名
func (ms *MultiSig) SignOne(tx *transaction.Transaction, inputIndex uint32, privateKey *ec.PrivateKey) (*[]byte, error) {
	if tx.Inputs[inputIndex].SourceTxOutput() == nil {
		return nil, transaction.ErrEmptyPreviousTx
	}

	// Print raw sighash preimage for debugging (before double SHA256)
	preimage, _ := tx.CalcInputPreimage(inputIndex, *ms.SigHashFlag)
	fmt.Printf("Go sighashData: %x\n", preimage)
	// Extract and print intermediate hashes
	if len(preimage) >= 68 {
		hashPrevouts := preimage[4:36]
		hashSequence := preimage[36:68]
		// hashOutputs is located 40 bytes from the end (4 locktime + 4 sighash flag + 32 hashOutputs)
		hashOutputs := preimage[len(preimage)-40-32 : len(preimage)-40]
		fmt.Printf("Go hashPrevouts: %x\n", hashPrevouts)
		fmt.Printf("Go hashSequence: %x\n", hashSequence)
		fmt.Printf("Go hashOutputs : %x\n", hashOutputs)
	}

	sh, err := tx.CalcInputSignatureHash(inputIndex, *ms.SigHashFlag)
	fmt.Printf("Go sighash: %x\n", sh)
	if err != nil {
		return nil, err
	}

	// Create unlocking script
	// s := script.NewFromBytes([]byte{})

	// Add OP_0 for the bug in CHECKMULTISIG
	// s.AppendOpcodes(script.Op0)

	// Sign with required number of private keys
	// for i := 0; i < ms.M; i++ {
	sig, err := privateKey.Sign(sh)
	fmt.Printf("GO privKey %x\n", privateKey.D.Bytes())
	fmt.Printf("GO r %x\n", sig.R.Bytes())
	fmt.Printf("GO s %x\n", sig.S.Bytes())
	if err != nil {
		return nil, err
	}

	sigBuf := make([]byte, 0)
	sigBuf = append(sigBuf, sig.Serialize()...)
	sigBuf = append(sigBuf, uint8(*ms.SigHashFlag))

	// if err = s.AppendPushData(sigBuf); err != nil {
	// 	return nil, err
	// }
	// }

	return &sigBuf, nil
}

// EstimateLength estimates the length of the unlocking script
func (ms *MultiSig) EstimateLength(_ *transaction.Transaction, _ uint32) uint32 {
	// OP_0 + M * (signature + sighash flag)
	return 1 + uint32(ms.M)*(71+1)
}

// FakeSign 创建一个假的签名脚本，方便计算长度
func FakeSign(m uint32) (*script.Script, error) {
	// 创建解锁脚本
	s := script.NewFromBytes([]byte{})

	// 添加 OP_0 用于修复 CHECKMULTISIG 中的 bug
	s.AppendOpcodes(script.Op0)

	// 创建假签名数据
	for i := 0; i < int(m); i++ {
		// 假签名数据：使用72字节的空数据模拟签名（最大DER签名长度）+ 1字节的空SigHashFlag
		fakeSigBuf := make([]byte, 72)
		// 添加一个空的SigHashFlag字节
		fakeSigBuf = append(fakeSigBuf, uint8(0))

		if err := s.AppendPushData(fakeSigBuf); err != nil {
			return nil, err
		}
	}

	return s, nil
}

func BuildSignScript(signs *[][]byte) (*script.Script, error) {
	// 创建解锁脚本
	s := script.NewFromBytes([]byte{})

	// 添加 OP_0 用于修复 CHECKMULTISIG 中的 bug
	s.AppendOpcodes(script.Op0)

	// 添加签名数据
	for _, sign := range *signs {
		sigBuf := make([]byte, 0)
		sigBuf = append(sigBuf, sign...)
		// sigBuf = append(sigBuf, uint8(*sigHashFlag))

		if err := s.AppendPushData(sigBuf); err != nil {
			return nil, err
		}

	}

	return s, nil
}
